templated AD and nonAD nodeElemConstraint

lynnmunday opened the pull request:
closes #30499

maxnezdyur committed:
quick_test with equal value

maxnezdyur committed:
update ad

maxnezdyur committed:
correct dofs for primary

lynnmunday committed:
adding generic AD nonAD base class for node elem constraints

lynnmunday committed:
this has a AD and nonAD that are then combined into a generic class.  It still
needs a non-templated base class fro nonlinearsystemBase and constraintWarehouse
to call.  This is following the pattern used in integratedBC

lynnmunday committed:
added a base class for nodeElem constraints that is called by constraintWarehouse
and nonlinearSystemBase.
This builds and all of the tests pass.

lynnmunday committed:
added test for AD equal value constraint.  Added errors and pure virual function to base class

lynnmunday committed:
moving common functions to base class.
Removing unused variables and functions.
Tests are now actually testing the AD output files.

closes #30499

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2868723/), step Clang format on 8a7883a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30502/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30502/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fc26d03b7a90fb8d0acc819941d48eb1f3ea17cb`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2868787/), step Clang format on f483ab0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30502/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30502/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fc26d03b7a90fb8d0acc819941d48eb1f3ea17cb`

maxnezdyur commented on framework/src/constraints/EqualValueEmbeddedConstraint.C:
```diff
@@ -41,20 +44,23 @@ EqualValueEmbeddedConstraint::validParams()
     _fe_problem(*parameters.get<FEProblem *>("_fe_problem")),
-    _formulation(getParam<MooseEnum>("formulation").getEnum<Formulation>()),
-    _penalty(getParam<Real>("penalty")),
+    _formulation(this->template getParam<MooseEnum>("formulation").template getEnum<Formulation>()),
```
Might have missed it, but for AD do you have an error for the kinematic formulation? Or am I wrong and the kinematic form works with AD?

lynnmunday replied:
It works with AD.  i'll add the ad kinematic tests.

lynnmunday replied:
I was wrong, AD with Kinematic constraints do not work.  The test is using PJFNK so it didn't need an accurate jacobian to pass.  I changed it to Newton and it fails.  I'll add an error for using kinematic with AD.

maxnezdyur reviewed (changes requested):
Have a few notes of things I'd like I'm not sure about. But in general, seems like this will make it easy to have templated nodeelem constaints for the future. Not for this PR, but I think this would be useful for other constraint classes too.

lynnmunday commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,62 @@
+  virtual ADReal computeQpResidual(Moose::ConstraintType type) = 0;
+
+  /// This is the virtual that derived classes should override for computing the Jacobian.
+  virtual Real computeQpJacobian(Moose::ConstraintJacobianType /*type*/) { return 0; }
+
+  /// This is the virtual that derived classes should override for computing the off-diag Jacobian.
+  virtual Real computeQpOffDiagJacobian(Moose::ConstraintJacobianType /*type*/,
+                                        unsigned int /*jvar*/)
+  {
+    return 0;
+  }
```
Yes, EqualValueEmbeddedConstraint won't build without this.  @GiudGiud said I would need some kind of partial specialization to only implement the nonAD version of these.

lynnmunday commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,62 @@
+  /// Computes the jacobian for the current element.
+  virtual void computeJacobian() override;
+
+  /// Computes d-residual / d-jvar...
+  virtual void computeOffDiagJacobian(unsigned int /*jvar*/) override {};
```
You're right, I removed it.

lynnmunday committed:
address review comments

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2869376/), step Docs: sync website on 15c8def wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30502/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2869391/), step Generate coverage on 15c8def wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>fc26d0</th>
      <th colspan=4>#30502 15c8de</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.41%</td>
      <td>-0.05%</td>
      <td>85.98%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>112542</td>
      <td>112484</td>
      <td>-58</td>
      <td>92</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19137</td>
      <td>19211</td>
      <td>+74</td>
      <td>15</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30502/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30502/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30502/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 85.98% is less than the suggested 90.0%



This comment will be updated on new commits.

lynnmunday committed:
disabled kinematic constraints when AD is used.

lynnmunday committed:
add documentation

GiudGiud commented on framework/doc/content/source/constraints/ADEqualValueEmbeddedConstraint.md:
```diff
@@ -0,0 +1,14 @@
+This version of `EqualValueEmbeddedConstraint` uses automatic differentiation, see [EqualValueEmbeddedConstraint.md] for information on formulation.
+
+!alert warning title=`ADEqualValueEmbeddedConstraint` only works for PENALTY formulation
+Automatic differentation cannot be used to compute Jacobian terms for [!param](/Constraints/ADEqualValueEmbeddedConstraint/formulation) = KINEMATIC because the KINEMATIC implementation enforces the constraint by enforcing the residual value on the secondary node prior to application of the constarint, $r_{s,copy}$ in [!eqref](EqualValueEmbeddedConstraint.md#eqn:kinematic).  This secondary node re ...
```
```suggestion
Automatic differentation cannot be used to compute Jacobian terms for [!param](/Constraints/ADEqualValueEmbeddedConstraint/formulation) = KINEMATIC because the KINEMATIC implementation enforces the constraint by enforcing the residual value on the secondary node prior to application of the constraint, $r_{s,copy}$ in [!eqref](EqualValueEmbeddedConstraint.md#eqn:kinematic).  This secondary node residual value is a nonAD copy of the residual and cannot be used to compute derivatives for the Jacobian.
```

GiudGiud commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,59 @@
+#include "NeighborCoupleableMooseVariableDependencyIntermediateInterface.h"
+
+/**
+ * A ADNodeElemConstraint is used when you need to create constraints between
```
```suggestion
 * An ADNodeElemConstraint is used when you need to create constraints between
```

GiudGiud commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,59 @@
+
+/**
+ * A ADNodeElemConstraint is used when you need to create constraints between
+ * a secondary node and a primary element. It works by allowing you to modify the
```
```suggestion
 * a subdomain of secondary nodes and a subdomain of primary elements. It works by allowing you to modify the
```

GiudGiud commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,59 @@
+/**
+ * A ADNodeElemConstraint is used when you need to create constraints between
+ * a secondary node and a primary element. It works by allowing you to modify the
+ * residual and jacobian entries on the secondary node and the primary element.
```
```suggestion
 * residual and jacobian entries on the secondary nodes and the primary elements.
```

GiudGiud commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,59 @@
+
+  ADNodeElemConstraint(const InputParameters & parameters);
+
+  /// Computes the residual Nodal residual.
```
?

GiudGiud commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,59 @@
+  /// prepare the _secondary_to_primary_map
+  virtual void prepareSecondaryToPrimaryMap() = 0;
+
+  /// This is the virtual that derived classes should override for computing the residual.
```
```suggestion
  /// This is the virtual method that derived classes should override for computing the residual.
```

GiudGiud commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,59 @@
+  virtual ADReal computeQpResidual(Moose::ConstraintType type) = 0;
+
+  /// This is the virtual that derived classes should override for computing the Jacobian.
+  virtual Real computeQpJacobian(Moose::ConstraintJacobianType /*type*/) { return 0; }
```
```suggestion
  virtual Real computeQpJacobian(Moose::ConstraintJacobianType /*type*/) { mooseAssert(false, "Should not be used"); }
```

GiudGiud commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,59 @@
+  virtual Real computeQpOffDiagJacobian(Moose::ConstraintJacobianType /*type*/,
+                                        unsigned int /*jvar*/)
+  {
+    return 0;
```
```suggestion
    mooseAssert(false, "Should not be used");
```

GiudGiud commented on framework/include/constraints/EqualValueEmbeddedConstraint.h:
```diff
@@ -54,9 +55,14 @@ class EqualValueEmbeddedConstraint : public NodeElemConstraint
   const enum class Formulation { KINEMATIC, PENALTY } _formulation;
   /// Penalty parameter used in constraint enforcement for kinematic and penalty formulations
-  const Real _penalty;
+  const GenericReal<is_ad> _penalty;
```
if ADEqual...Constaint does not work for kinematic, isnt this always a constant?

lynnmunday replied:
penalty is constant.  NonAD will work for kinematic, AD won't.

GiudGiud commented on framework/include/constraints/NodeElemConstraintBase.h:
```diff
@@ -0,0 +1,121 @@
+}
+
+/**
+ * A NodeElemConstraintBase is used when you need to create constraints between
+ * a secondary node and a primary element. It works by allowing you to modify the
+ * residual and jacobian entries on the secondary node and the primary element.
```
```suggestion
 * A NodeElemConstraintBase is a base class for constraints between a subdomain with
 * secondary nodes and a subdomain with primary elements. It works by allowing you to modify the
 * residual and jacobian entries on the secondary nodes and the primary elements.
```

GiudGiud commented on framework/include/constraints/NodeElemConstraintBase.h:
```diff
@@ -0,0 +1,121 @@
+   * the residual previously at that node for that variable.
+   */
+  bool _overwrite_secondary_residual;
+  // fixme move
```
let s move it

GiudGiud commented on framework/include/constraints/NodeElemConstraintBase.h:
```diff
@@ -0,0 +1,121 @@
+  bool _overwrite_secondary_residual;
+  // fixme move
+public:
+  SparseMatrix<Number> * _jacobian;
```
docstirng

GiudGiud commented on framework/include/constraints/NodeElemConstraintBase.h:
```diff
@@ -0,0 +1,121 @@
+  /// Side test function
+  const VariableTestValue & _test_primary;
+
+  const std::map<dof_id_type, std::vector<dof_id_type>> & _node_to_elem_map;
```
docstring
is this a local one or is this the one from the MooseMesh

lynnmunday replied:
MooseMesh

GiudGiud commented on framework/include/constraints/NodeElemConstraintBase.h:
```diff
@@ -0,0 +1,121 @@
+   * @return bool When this returns true the secondary's residual as computed by the constraint will
+   * _replace_ the residual previously at that node for that variable.
+   */
+  virtual bool overwriteSecondaryResidual();
```
```suggestion
  virtual bool overwriteSecondaryResidual() const;
```

GiudGiud commented on framework/include/constraints/NodeElemConstraintBase.h:
```diff
@@ -0,0 +1,121 @@
+   * The variable on the primary elem.
+   * @return MooseVariable & a reference to the primary variable
+   */
+  virtual MooseVariable & primaryVariable() { return _primary_var; }
```
```suggestion
  virtual const MooseVariable & primaryVariable() const { return _primary_var; }
```

GiudGiud commented on framework/include/constraints/NodeElemConstraintBase.h:
```diff
@@ -0,0 +1,121 @@
+  /// Compute the value the secondary node should have at the beginning of a timestep.
+  virtual Real computeQpSecondaryValue() = 0;
+
+  MooseVariable & _var;
```
docstring

GiudGiud commented on framework/include/constraints/NodeElemConstraintBase.h:
```diff
@@ -0,0 +1,121 @@
+
+  /// current node being processed
+  const Node * const & _current_node;
+  const Elem * const & _current_elem;
```
```suggestion
  /// current element being processed
  const Elem * const & _current_elem;
```

GiudGiud commented on framework/src/constraints/ADNodeElemConstraint.C:
```diff
@@ -0,0 +1,78 @@
+ADNodeElemConstraint::validParams()
+{
+  InputParameters params = NodeElemConstraintBase::validParams();
+  return params;
```
missing class description?

lynnmunday replied:
Do I need one if this class can't be in the input file?

GiudGiud replied:
ah no you're good

GiudGiud commented on framework/src/constraints/ADNodeElemConstraint.C:
```diff
@@ -0,0 +1,78 @@
+  for (_i = 0; _i < _test_primary.size(); _i++)
+    primary_residual[_i] += computeQpResidual(Moose::Primary);
+
+  addJacobian(_assembly, primary_residual, _primary_var.dofIndicesNeighbor(), _var.scalingFactor());
```
which equation is it? it s the scaling factor for that equation

since it s the dofs of the primary, should be the scaling factor for the primary

maxnezdyur replied:
This is how the contributions are added before the AD change. It uses the _var for both. But I don't know if that's incorrect.
https://github.com/idaholab/moose/blob/122840c1fdf80ef78a179bb019702c45924ec46d/framework/src/systems/NonlinearSystemBase.C#L2601-L2613

GiudGiud replied:
it's wrong there too imo.
And i see the same-ish error 300 lines above for NodeFaceConstraints.

@lindsayad thoughts? Are we wrong?

lindsayad replied:
No, you're not wrong. Looks like a bug to me for sure

lynnmunday replied:
https://github.com/idaholab/moose/issues/30530

I made a two variable test and it doesn't work for this constraint so I added an error to this PR and an issue:  https://github.com/idaholab/moose/issues/30530

I didn't try nodeFaceConsraint

GiudGiud replied:
@lynnmunday can you please also fix the mistake in the NonlinearSystem that Max pointed out? and same mistake about 300 lines above for NodeFace constratins

lynnmunday replied:
@GiudGiud I changed those two lines in NonlinearSystem base but they do not fix the convergence problem in issue #30530    

I did add scaling to the two variable nodeElem constraint problem and it will fail without fixing the scaling in ADNodeElemConstraint so that is working.

GiudGiud commented on framework/src/constraints/NodeElemConstraintBase.C:
```diff
@@ -0,0 +1,76 @@
+}
+
+bool
+NodeElemConstraintBase::overwriteSecondaryResidual()
```
```suggestion
NodeElemConstraintBase::overwriteSecondaryResidual() const
```

GiudGiud reviewed (changes requested):
most comments apply to things prior to this PR but it would be nice to get this class up to code

lynnmunday committed:
added test for two different variables on different domains.  This does not work for the nonAD version of NodeElemConstraint so I added an error if the variables are different and tested the error.  It does work for the AD version so I added a test for that.

lynnmunday committed:
address reviewer comments and remove some unneeded includes

Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

fixing constexpr if statement

lynnmunday committed:
adding variable scaling to the two variable test.
This will fail without using the correct primary_var scalingFactor for the primary variable Jacobian terms.

Comments about NonlinearSystemBase changes:
The changes to nonlinear system base are untested but seem correct.
The nodeElem change requires a nonAD node elem test with two variables which is currently disabled
and it still fails to converge with this change.
This change didn't change the convergence at all for that case.

moosebuild commented:
Job [Python 3.9](https://civet.inl.gov/job/2875917/) on 0e70d96 : invalidated by @lynnmunday

lynnmunday commented:
@GiudGiud I addressed all of the comments in your review.  Could you take another look and merge if its good?

GiudGiud commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,56 @@
+
+  ADNodeElemConstraint(const InputParameters & parameters);
+
+  /// Computes the Nodal residual.
+  virtual void computeResidual() override;
+
+  /// Computes the jacobian for the current element.
+  virtual void computeJacobian() override;
```
can we check this
Nodal residual and element Jacobian?

maxnezdyur reviewed (approved):
Fixed all my comments. Like I said in one of the comments, would be nice to get this done for the other constraints but now we have a model of how to do that efficiently!

GiudGiud commented on framework/include/constraints/ADNodeElemConstraint.h:
```diff
@@ -0,0 +1,56 @@
+  virtual void computeJacobian() override;
+
+protected:
+  /// prepare the _secondary_to_primary_map
```
```suggestion
  /// prepare the _secondary_to_primary_map (see NodeElemConstraintBase)
```

GiudGiud commented on framework/include/constraints/NodeElemConstraintBase.h:
```diff
@@ -0,0 +1,125 @@
+
+  /// Shape function on the secondary side.
+  VariablePhiValue _phi_secondary;
+  /// Shape function on the secondary side.  This will always only have one entry and that entry will always be "1"
```
I think this is only true for lagrange variables?

lynnmunday replied:
It is the shape function for a single node that is the secondary "surface".  It is not element connected to the node.

GiudGiud replied:
yes but that shape function is set to be 1 because that's the value of the lagrange shape function at that node

lynnmunday replied:
This is hardcoding the shape function for a node.  I think this has more to do with setting the secondary_primary lagrange multiplier than the type of shape function.  I don't think this is right for elements with dofs that aren't located at physical coordinates, I don't know what that would do.  This was already in the code before I started this PR to add AD so this is the behavior people are currently using.

GiudGiud replied:
yeah I m thinking we should be limiting the FE families allowed for the variables here. ANother PR

GiudGiud commented on framework/src/constraints/EqualValueEmbeddedConstraint.C:
```diff
@@ -41,20 +43,25 @@ EqualValueEmbeddedConstraint::validParams()
 {
   _overwrite_secondary_residual = false;
   prepareSecondaryToPrimaryMap();
+  if (is_ad && (_formulation == Formulation::KINEMATIC))
```
```suggestion
  if (constexpr is_ad && (_formulation == Formulation::KINEMATIC))
```

GiudGiud commented on test/tests/constraints/equal_value_embedded_constraint/tests:
```diff
@@ -119,8 +110,95 @@
+  []
+  [error_two_var_AD]
+    design = 'source/constraints/ADEqualValueEmbeddedConstraint.md'
+    requirement = "The system shall error when the nonAD constraint is used for coupling two variables."
```
does it converge with PJFNK ? is the Jacobian is not 100% correct witohut AD, makes sense that it does not converge with Newton

lynnmunday replied:
I didn't try PJFNK.  The Jacobian is wrong but I didn't look at how wrong.  The residual diverges.

GiudGiud replied:
Let's try it. These constraints were meant to be used with PJFNK since the Jacobians were not correct

lynnmunday replied:
We had assumed the nonAD version of these constraints had a correct Jacobian.  We didn't know it was incorrect until I added the test in this PR that coupled two different variables.  The nonAD version of this constraint works with Newton when this constraint is coupling the same variable.

GiudGiud replied:
>  It does not work for kinematic and PJFNK

It worked fine on the test case. you just need to adjust the scaling.
But you're right, the jacobian is terribly wrong. 

We need an issue for this. 
I m fine with the error

lynnmunday replied:
You are right, I purposefully put in bad manual scaling to highlight the issue.  It would converge without the bad scaling.  I added an issue that links to this PR and test.

GiudGiud replied:
ok found one error which prevents getting the correct Jacobian
for off-diag on primary side, we are actually using the wrong dofs (sending to secondary side)
should be
```

                  // Cache the jacobian block for the primary side
                  nec->addJacobian(_fe_problem.assembly(0, number()),
                                   nec->_Kne,
                                   nec->primaryVariable().dofIndicesNeighbor(),
                                   nec->_connected_dof_indices,
                                   nec->primaryVariable().scalingFactor());
```

but it s not the only problem.
the getConnectedDofs is not finding any connected DOFs for the primary variables

lynnmunday replied:
how did you find that, are stepping through this in gdb?

GiudGiud replied:
the node_to_elem maps, is not going to get elements that live on the primary side? 
I think we need more logic in getConnectedDofs to go grab those

lindsayad replied:
why is this test problem so big 😦

lynnmunday commented:
@GiudGiud Its ready for another review.

lynnmunday commented:
@GiudGiud I addressed all of your comments, could you merge this?

GiudGiud commented on test/tests/constraints/equal_value_embedded_constraint/tests:
```diff
@@ -119,8 +110,95 @@
+  [error_two_var_AD]
+    design = 'source/constraints/ADEqualValueEmbeddedConstraint.md'
+    requirement = "The system shall error when the nonAD constraint is used for coupling two variables."
+    issues = '#30499'
```
@lynnmunday just adding the new issue there is the last item

GiudGiud replied:
done

GiudGiud committed:
Add issue to test showing non-AD equal value constraint 2 variable problem
reword requirement

GiudGiud reviewed (approved):
Thanks for the contribution!

hugary1995 commented:
Could you leave me with one more day to review this? If I don't get back just ignore me.

lindsayad committed:
Fix ElementNeighbor Jacobian for different secondary-primary vars

lindsayad committed:
Update nonAD test to RunApp cuz it works now

Closes #30530

GiudGiud commented:
sure! no rush

lindsayad commented:
@GiudGiud I'm pretty sure you just overwrote some of @lynnmunday's commits

lindsayad commented:
It looks like you dropped

* lynnmunday/moose@5228f6326d 2025-05-20 Lynn Munday fixing constexpr if statement
* lynnmunday/moose@63d53b18cb 2025-05-20 Lynn Munday Apply suggestions from code review

GiudGiud commented:
I squashed the "address review comment" and the "fix constexpr" ones with the commit they pertained to

hugary1995 commented:
I have a high-level question before I start my review. What's the motivation of the AD constraint?

By reading the issue description, it seems to me that the motivation was that the non-AD version has wrong Jacobian if we try to tie two different variables. But in the last commits by @lindsayad , it seems like that problem has been resolved -- correct me if I interpreted it wrong.

In this comment https://github.com/idaholab/moose/pull/30502#issuecomment-2894659091, @GiudGiud mentioned that AD might be useful for complicated cases. But I fail to come up with such a case where the non-AD constraint would not work (assuming its Jacobian has been fixed). The variables being constrained are independent, nonlinear variables, so I don't see much value in adding an AD version here. Am I missing something?

moosebuild commented:
Job [Framework 1](https://civet.inl.gov/job/2888024/) on df16fe8 : invalidated by @lynnmunday

hugary1995 commented:
I see. There you compute the bond stress for the constraint, which may have dual numbers from arbitrary variables. I agree AD would be useful then. Let me finish my review real quick.

hugary1995 commented:
My elem-elem constraint stuff is just another way of enforcing the constraint. I haven't implemented bond-slip constraint using it. Speaking of Jacobian, I expect it to face the exact same issues as those you are facing here.

lynnmunday commented:
@GiudGiud and @lindsayad Thanks for fixing this.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2888045/), step Verify coverage on df16fe8 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2888045/) on df16fe8 : invalidated by @lynnmunday

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2888045/), step Verify coverage on df16fe8 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

lynnmunday commented:
@GiudGiud Can you merge this?
